abci/multiplexer.go

```diff
@@ -46,42 +82,132 @@ func NewVersions(v ...Version) (Versions, error) {
 }
 
 // NewMultiplexer creates a new ABCI wrapper for multiplexing
-func NewMultiplexer(
-	logger log.Logger,
-	v *viper.Viper,
-	latestApp servertypes.ABCI,
-	versions Versions,
-	chainID string,
-	applicationVersion uint64,
-) (proxy.ClientCreator, func() error, error) {
-	var noOpCleanUp = func() error {
-		return nil
-	}
-
+func NewMultiplexer(svrCtx *server.Context, svrCfg serverconfig.Config, clientCtx client.Context, appCreator servertypes.AppCreator, versions Versions, chainID string, applicationVersion uint64) (*Multiplexer, error) {
 	if err := versions.Validate(); err != nil {
-		return nil, noOpCleanUp, fmt.Errorf("invalid versions: %w", err)
+		return nil, fmt.Errorf("invalid versions: %w", err)
 	}
 
-	wrapper := &Multiplexer{
-		logger:         logger,
-		latestApp:      latestApp,
+	mp := &Multiplexer{
+		svrCtx:         svrCtx,
+		svrCfg:         svrCfg,
+		clientContext:  clientCtx,
+		appCreator:     appCreator,
+		logger:         svrCtx.Logger.With("multiplexer"),
+		latestApp:      nil, // app will be initialized if required by the multiplexer.
 		versions:       versions,
 		chainID:        chainID,
 		lastAppVersion: applicationVersion,
+		cleanupFns:     make([]func() error, 0),
+	}
+
+	return mp, nil
+}
+
+// isGrpcOnly checks if the GRPC-only mode is enabled using the configuration flag.
+func (m *Multiplexer) isGrpcOnly() bool {
+	return m.svrCtx.Viper.GetBool(flagGRPCOnly)
+}
+
+// registerCleanupFn enables the registration of additional cleanup functions that get called during Cleanup
+func (m *Multiplexer) registerCleanupFn(cleanUpFn func() error) {
+	m.cleanupFns = append(m.cleanupFns, cleanUpFn)
+}
+
+func (m *Multiplexer) Start() error {
+	m.g, m.ctx = getCtx(m.svrCtx, true)
+
+	emitServerInfoMetrics()
+
+	// startApp starts the underlying application, either native or embedded.
+	if err := m.startApp(); err != nil {
+		return err
 	}
 
+	if m.isGrpcOnly() {
+		m.logger.Info("starting node in gRPC only mode; CometBFT is disabled")
+		m.svrCfg.GRPC.Enable = true
+	} else {
+		if m.tmNode == nil {
+			m.logger.Info("starting comet node")
+			if err := m.startCmtNode(); err != nil {
+				return err
+			}
+		}
+	}
+
+	if m.latestApp == nil {
+		m.logger.Info("using embedded app, not continuing with grpc or api servers")
+		return m.g.Wait()
+	}
+
+	if err := m.enableGRPCAndAPIServers(m.latestApp); err != nil {
+		return err
+	}
+
+	// wait for signal capture and gracefully return
+	// we are guaranteed to be waiting for the "ListenForQuitSignals" goroutine.
+	return m.g.Wait()
+}
+
+// enableGRPCAndAPIServers enables the gRPC and API servers for the provided application if configured to do so.
+// It registers transaction, Tendermint, and node services, and starts the gRPC and API servers if enabled.
+func (m *Multiplexer) enableGRPCAndAPIServers(app servertypes.Application) error {
+	if app == nil {
```

Comment:
all this logic is adapted from what was already in start.go, just slightly modified now that it is on a type instead of a standalone function.